icu4j/main/classes/core/src/com/ibm/icu/message2/Mf2DataModelFormatter.java

```diff
@@ -0,0 +1,277 @@
+                    selectorFunctions.add(funcFactory.createSelector(locale, opt));
+                } else {
+                    throw new IllegalArgumentException("Unknown selector type: " + functionName);
+                }
+            }
+            // This should not be possible, we added one function for each selector, or we have thrown an exception.
+            // But just in case someone removes the throw above?
+            if (selectorFunctions.size() != selectors.size()) {
+                throw new IllegalArgumentException("Something went wrong, not enough selector functions, "
+                        + selectorFunctions.size() + " vs. " + selectors.size());
+            }
+
+            patternToRender = null;
+            // Iterate "vertically", through all variants
+            for (Entry<SelectorKeys, Pattern> variant : dm.getVariants().entrySet()) {
+                int maxCount = selectors.size();
+                List<String> keysToCheck = variant.getKey().getKeys();
+                if (selectors.size() != keysToCheck.size()) {
+                    throw new IllegalArgumentException("Mismatch between the number of selectors and the number of keys: "
+                            + selectors.size() + " vs. " + keysToCheck.size());
+                }
+                boolean matches = true;
+                // Iterate "horizontally", through all matching functions and keys
+                for (int i = 0; i < maxCount; i++) {
+                    Expression selector = selectors.get(i);
+                    String valToCheck = keysToCheck.get(i);
+                    Selector func = selectorFunctions.get(i);
+                    Map<String, Object> options = mf2OptToFixedOptions(selector.getOptions());
+                    if (!func.matches(variableToObjectEx(selector.getOperand(), arguments), valToCheck, options)) {
+                        matches = false;
+                        break;
+                    }
+                }
+                if (matches) {
+                    patternToRender = variant.getValue();
+                    break;
+                }
+            }
+        }
+
+        // TODO: check that there was an entry with all the keys set to `*`
+        if (patternToRender == null) {
+            // If there was a case with all entries in the keys `*` this should not happen
+            throw new IllegalArgumentException("The selection went wrong, cannot select any option.");
+        }
+
+        StringBuilder result = new StringBuilder();
+        for (Part part : patternToRender.getParts()) {
+            if (part instanceof Text) {
+                result.append(part);
+            } else if (part instanceof Expression) { // Placeholder is an Expression
+                FormattedPlaceholder fr = formatPlaceholder((Expression) part, arguments, false);
+                FormattedValue fv = fr.getFormattedValue();
+                if (fr.getInput() == null && (fv == null || fv.toString() == null)) {
```

Comment:
fv cannot be null any more; and why would its .toString() be null?

Reply:
Added TODO(before 72)

Reply:
Done